llvm/lib/Target/AIE/ReservedRegsLICM.cpp

```diff
@@ -0,0 +1,387 @@
+      return &Candidates.try_emplace(Reg, Reg).first->second;
+    }
+  }
+  return nullptr;
+}
+
+class ReservedRegsLICM : public MachineFunctionPass {
+  const TargetRegisterInfo *TRI;
+  MachineRegisterInfo *MRI;
+
+  /// Whether the MachineFunction got changed
+  bool Changed = false;
+
+public:
+  static char ID;
+  ReservedRegsLICM() : MachineFunctionPass(ID) {}
+
+  bool runOnMachineFunction(MachineFunction &MF) override;
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.addRequired<MachineLoopInfo>();
+    AU.addPreserved<MachineLoopInfo>();
+    MachineFunctionPass::getAnalysisUsage(AU);
+  }
+
+private:
+  /// Collect all simplifiable reserved registers that all livein for \p L.
+  BitVector collectLoopReservedLiveins(const MachineLoop &L);
+
+  /// Go through \p L to look for instructions to hoist into the preheader or
+  /// sink into the exit block.
+  void runOnLoop(MachineLoop &L);
+
+  /// Find instructions to sink into the exit block
+  void processForExitSink(MachineLoop &L, const BitVector &ReservedLiveins);
+
+  /// Sink \p Cand to \p L's exit block if it is safe to do so.
+  bool trySinkToExitBlock(const CandidateInfo &Cand, MachineLoop &L);
+
+  /// Find instructions to hoist to the preheader
+  void processForPreheaderHoist(MachineLoop &L,
+                                const BitVector &ReservedLiveins);
+
+  /// Hoist \p Cand to \p L's preheader if it is safe to do so.
+  bool tryHoistToPreHeader(const CandidateInfo &Cand, MachineLoop &L);
+
+  /// Verify if \p Cand is loop invariant and can be safely hoisted.
+  /// \pre Cand->DefinedReg has a unique live value within the loop. This is
+  /// verified by processForExitSink() or processForPreheaderHoist().
+  bool isLoopInvariantInst(const CandidateInfo &Cand, const MachineLoop &L);
+};
+
+} // end anonymous namespace
+
+char ReservedRegsLICM::ID;
+
+char &llvm::ReservedRegsLICMID = ReservedRegsLICM::ID;
+
+INITIALIZE_PASS_BEGIN(ReservedRegsLICM, DEBUG_TYPE,
+                      "Machine LICM for reserved regs", false, false)
+INITIALIZE_PASS_DEPENDENCY(MachineLoopInfo)
+INITIALIZE_PASS_END(ReservedRegsLICM, DEBUG_TYPE,
+                    "Machine LICM for reserved regs", false, false)
+
+llvm::FunctionPass *llvm::createReservedRegsLICMPass() {
+  return new ReservedRegsLICM();
+}
+
+bool ReservedRegsLICM::runOnMachineFunction(MachineFunction &MF) {
+  if (skipFunction(MF.getFunction()))
+    return false;
+
+  const TargetSubtargetInfo &ST = MF.getSubtarget();
+  TRI = ST.getRegisterInfo();
+  MRI = &MF.getRegInfo();
+  assert(MRI->isSSA() && "ReservedRegsLICM can only deal with SSA vregs");
+
+  LLVM_DEBUG(dbgs() << "******** Reserved register LICM: " << MF.getName()
+                    << " ********\n");
+
+  MachineLoopInfo &MLI = getAnalysis<MachineLoopInfo>();
+
+  SmallVector<MachineLoop *, 4> Loops = MLI.getBase().getLoopsInPreorder();
+  for (MachineLoop *L : reverse(Loops)) {
```

Comment:
nit: we only do leaf loops. reverse is not really necessary.

Reply:
We do check all loops now, `getLoopsInPreorder` will find nested ones as well.